### PR TITLE
Add test for duplicate events with nested dispatch

### LINKED ISSUE #8559: Click handler can get called twice in "non-prod" mode.
I think this is a bug that only happens in "non-prod" mode.

I noticed this using the react-dropzone library but I think I found a simple use case to reproduce the problem.  If you check out the [fiddle](https://jsfiddle.net/69z2wepo/64978/) you can see that click handler is called twice even though the element is only clicked once.  Here is an example showing it ["working and not working"](https://jsfiddle.net/f2ntq2h6/1/)  

I believe it has something to do with the outer div's click listener and this chunk of code:

```js
if (true) {
  /**
   * To help development we can get better devtools integration by simulating a
   * real browser event.
   */
  if (typeof window !== 'undefined' && typeof window.dispatchEvent === 'function' && typeof document !== 'undefined' && typeof document.createEvent === 'function') {
    var fakeNode = document.createElement('react');
    ReactErrorUtils.invokeGuardedCallback = function (name, func, a) {
      var boundFunc = func.bind(null, a);
      var evtType = 'react-' + name;
      fakeNode.addEventListener(evtType, boundFunc, false);
      var evt = document.createEvent('Event');
      // $FlowFixMe https://github.com/facebook/flow/issues/2336
      evt.initEvent(evtType, false, false);
      fakeNode.dispatchEvent(evt);
      fakeNode.removeEventListener(evtType, boundFunc, false);
    };
  }
}
```

I tested this on Chrome/Mac I have also seen this behavior on React 14.8.

### PR DESCRIPTION
**Pull request by Andarist**
Still have tests to write, but they should be pretty straightforward after PR aproval.

Also had no idea where I should put `once` utility function, so its place in the PR for now is just a temporary.

This PR is just a proposal of a fix, the thing I've come with to tackle this problem, I'm open to change the code appropriately if somebody comes up with the better idea how to solve this.

cc @Aweary

### CODE REVIEW

**Comment by Aweary** [comment]
Thanks for the PR @Andarist, and for the write up in #8559. I'm wondering if we can safely just get rid of this whole dev-only abstraction. We use it elsewhere (like for [`componentWillUnmount`](https://github.com/facebook/react/blob/7ef856aa3676f898ac5014f77a3e176bdf727350/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L574)) so it seems kind of strange for it to be dispatching an event for that.

**Comment by gaearon** [comment]
It dispatches a fake event so that errors originating in user code are on the top of the stack and "break on uncaught exceptions" works. I think that's the main reason.

**Comment by Andarist** [comment]
Hm, its in fact kinda weird that `componentWillUnmount` can use this __DEV__ version `ReactErrorUtils.invokeGuardedCallback`.

It seems that it should always (regardless the env) use this one - https://github.com/facebook/react/blob/7ef856aa3676f898ac5014f77a3e176bdf727350/src/renderers/shared/utils/ReactErrorUtils.js#L24 .

Maybe that should be fixed too?

**Comment by Aweary** [comment]
We can update that call in `unmountComponent` to use `invokeGuardedCallbackWithCatch` which is the `try` / `catch` version in all environments.

As for using `once` here, it feels like we're addressing the symptom instead of the issue. Can we deregister the event handler in the bound function, before we call `func`?  That way the fake event handler won't be called again in the first place.

**Comment by Andarist** [comment]
So I've pushed out the new version. It might seem that it contains some unnecessary changes, but:
1. it seems that `jest.resetModuleRegistry` is a deprecated or internal API (not in the docs) and it only calls documented - `jest.resetModules`
2. I've added `jest.resetModules()` to the `beforeEach`, because of the jest caching I've stumbled upon an issue with `ReactErrorUtils` being evaluated just once (and second force evaluation in other production-only test) per this file. So when I've put my new test after the one which forced `ReactErrorUtils` to be evaluated as in production mode, it has stayed the same (no re-evaluation within `beforeEach`) - therefore I couldnt test properly and wasted some time on figuring this out. I could change the test order, but thought this change might save somebody's time in the future. WDYT?

Also - the test's comment might not be the best, but cannot express it better atm

File: src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js [hunk 1]
```diff
@@ -63,13 +64,23 @@ describe('ReactErrorUtils', () => {
...
+
+    it('shouldn\'t call the callback again while executing synchronously induced other call in development', () => {
+      var callback = jest.fn(() => {
+        ReactErrorUtils.invokeGuardedCallback('foo', callback2, 'arg', true);
+      });
+      var callback2 = jest.fn();
+      ReactErrorUtils.invokeGuardedCallback('foo', callback, 'arg', true);
+      expect(callback).toHaveBeenCalledTimes(1);
+      expect(callback2).toHaveBeenCalledTimes(1);
+    });
```
Comment (Aweary) [inline src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js]: Does this test fail if you revert the changes in `ReactErrorUtils`?

**Comment by Andarist** [comment]
Yes. But test file would need to stay as in my PR so the ordering issue is
solved by resetModules

**Comment by Andarist** [comment]
CicrcleCI didnt pass on 1 container out of 4, but aint sure why, as locally I do not have any problems and 3 containers passed the tests just fine.

Did you start using Circle recently? I dont recall it being used before.

File: src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js (hunk 1 above)
Comment (Andarist) [inline src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js]: Yes. But test file would need to stay as in my PR so the ordering issue is solved by `resetModules`

**Comment by Andarist** [comment]
@aweary 
i've seen that you have closed and reopened the issue - probably to force CircleCI to run the build

However it didnt pass on some of the containers (similarly like before). After that I've rebased my branch which forced another build - and the same thing again.

Do you have any idea why this might happen? My changes were really minor and I dont see why this is happening.

**Comment by aweary** [comment]
@Andarist don't worry, it's failing for unrelated reasons (see https://github.com/facebook/react/commit/378ef5e7300d3c6c1ea10b4fcf555611c450f8cd)

**Comment by gaearon** [comment]
(You should be able to rebase to fix this?)

**Comment by Andarist** [comment]
I've done that, but 1 container is still failing.

I think it's connected with ongoing fiber development and the fact I've added a test case. From what we can see overall in https://circleci.com/gh/facebook/react/760#tests/containers/2 `./scripts/circleci/test_fiber.sh` is not exactly passing - just like on my branch here - https://circleci.com/gh/facebook/react/772#tests/containers/2 , but the outcome is different, master is OK, but my branch is not.

Just a wild idea (cause im not aware how the things are developed atm) - should i run `scripts/fiber/record-tests` now and add it to the commit? Does it make a difference?

**Comment by gaearon** [comment]
>Just a wild idea (cause im not aware how the things are developed atm) - should i run scripts/fiber/record-tests now and add it to the commit? Does it make a difference?

Yes, exactly! Sorry it's not documented.

**Comment by Andarist** [comment]
Rebased + ammended with `scripts/fiber/tests-passing.txt`. Working like a charm and PR is green ;)

File: src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js [hunk 2]
```diff
@@ -71,5 +72,15 @@ describe('ReactErrorUtils', () => {
       __DEV__ = true;
       global.process = oldProcess;
     });
+
+    it('shouldn\'t call the callback again while executing synchronously induced other call in development', () => {
```
Comment (gaearon) [inline src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js]: Is there any way to write a test for observable behavior instead? That is, write a test that relies on `ReactDOM` and `React`, but not on `ReactErrorUtils` itself. Otherwise, it's hard to tell what exactly this is fixing.

Comment (Andarist) [inline src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js]: Sure thing, should the new one be located in the same file or should I move it somewhere else? gonna prepare better test later 2day

Comment (gaearon) [inline src/renderers/shared/utils/__tests__/ReactErrorUtils-test.js]: `ReactDOM-test` might be an okay place to put it.

**Comment by Andarist** [comment]
Ok, so I've changed the test (removed old one too), so it should clearly indicate what's it testing against.

I had few issues with understanding how should I handle it and if any of you could clarify things a little bit it would be great.

First of all it is totally unclear for me how to cause an event, there are numerous ways to do so and had no idea which to chose, finally just have chosen the one which gave me my expected results, but I had to test a few. Anyway - not much of a problem, but overally such things could be described somewhere (or maybe ive just didnt find this?)

Also when during my experiments I've put:
```javascript
beforeEach(() => {
  jest.resetModules();
  ReactDOM = require('ReactDOM');
});
```
it has stopped to work and I have no idea why, wasted some time on this one as in [`executeDispatchesInOrder`](https://github.com/facebook/react/blob/5659bd2a6311e70e8aed97e33d92b84e57dd15ca/src/renderers/shared/shared/event/EventPluginUtils.js#L107) `dispatchListeners` were falsy and no handler was called. Removing the `beforeEach` fixed this.

But the biggest obstacle Ive met is that I think that code during testing is going through some other path than in the latest React. There are some fiber-called things in the call stack which are absent in production code. I think this might be responsible for the different behaviour that I have wanted to debug - but ultimately failed. It might (or not) be relevant for your guys though.

The problem was that the real life situation was acting like this before the fix - handler1 > handler1 > handler2. But during tests it was handler1 > handler1 > handler1 > handler1 > ..., so effectively it created an endless loop.

My fix is working for both scenarios, but I was curious about this difference so I wanted to investigate. I THINK its caused by some logic connected to [`enqueueEvents`](https://github.com/facebook/react/blob/5659bd2a6311e70e8aed97e33d92b84e57dd15ca/src/renderers/shared/shared/event/EventPluginHub.js#L196) - with each iteration of the dispatch loop the first handler somehow go into being accumulated there again and again and again, which was not the case with real life scenario. Hope it means something more for you than to me :)

If somebody is interested I have a call stack for this endless loop which I can post here later.

**Comment by Andarist** [comment]
@aweary 
#8961 fixed this, also confirmed that just before #8961 the issue still occured

That being said v15 branch wont ever have this fix from #8961 as its fiber-related heavily. Dont know how you guys approach this kind of situations and if you diverge things for 2 different major releases.

Also the test could be incorporated into the v16 to ensure regression wont happen here in the future.

Let me know what I should do about this PR, i.e I can split this into 2 commits - 1st for v15 fix, 2nd for test, so only the latter can be cherry-picked for master, while both could be merged to v15 line.

**Comment by aweary** [comment]
@Andarist if https://github.com/facebook/react/pull/8961 fixed this then we should be good to go, we have a release manager that will deal with this and make sure changes to utilities used by `15` are included in the next release 👍

**Comment by Andarist** [comment]
@aweary How `v15` can get this fix from #8961 ? From what I see this PR is Fiber-related, doesnt it imply this is for `v16+` only?

Also what about the test? Both versions could benefit from it, guarding against regression in this matter.

**Comment by aweary** [comment]
`ReactErrorUtils.js` is a shared module so we should be able to integrate those changes into the stack renderer as well. Looking at #8961 the changes made to `ReactErrorUtils` don't appear incompatible with `15`. But actually I think this next `15` release has to be done manually, maybe @gaearon can clarify how we'd approach this

>Also what about the test? Both versions could benefit from it, guarding against regression in this matter.

Yeah definitely, I'll re-open if you can just remove the changes to `ReactErrorUtils` and just include the test

**Comment by Andarist** [comment]
@aweary 
So I've rebased and ammended my change. Besides the test guarding against future regressions Ive also decided to apply also a change from original PR.

I've moved unsubscribe logic to the start of the event's callback [here](https://github.com/Andarist/react/blob/7af2299a99db5a105b6e6909510673d49417ae69/src/renderers/shared/utils/ReactErrorUtils.js#L58). I think it wont harm anything as tests are passing and it serves as additional guard - the only reason that the other refactor has fixed this issue is the fact that `depth` was added to the event type. Without that it would still be broken, so future changes might break it.

With this line moved to the callback, it stays double guarded by different event types + by clean up logic moved. Also it will ensure that cleaning up will actually happen in case of an error and we wont end up with stale listener.

Possibly even with such stale listener we might end up with other bugs, when some other code would dispatch new event with the same name and depth. I need to check this though and possibly write tests for such case.

**Comment by Andarist** [comment]
@aweary ping ;)

this is a silly PR really, but would like to have it stop hanging as open in my PR tab, I understand though if you are too busy right now

**Comment by aweary** [comment]
@Andarist I think we can just merge this since it's just adding additional test coverage now. Thanks for this, sorry about the huge delay 😄

**Comment by Andarist** [comment]
No problem, sorry that i forgot to adjust commit message while rebasing
